Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,275 @@
+# 
+#  Authors:
+#  - CIL Developers, listed at: https://github.com/TomographicImaging/CIL/blob/master/NOTICE.txt
+#  - Daniel Deidda (National Physical Laboratory, UK)
+#  - Claire Delplancke (Electricite de France, Research and Development)
+#  - Ashley Gillman (Australian e-Health Res. Ctr., CSIRO, Brisbane, Queensland, Australia)
+#  - Zeljko Kereta (Department of Computer Science, University College London, UK)
+#  - Evgueni Ovtchinnikov (STFC - UKRI)
+#  - Georg Schramm (Department of Imaging and Pathology, Division of Nuclear Medicine, KU Leuven, Leuven, Belgium)
+
+
+from .ApproximateGradientSumFunction import ApproximateGradientSumFunction
+import numpy as np
+import numbers
+
+
+class SVRGFunction(ApproximateGradientSumFunction):
+
+    r"""
+    The Stochastic Variance Reduced Gradient (SVRG) function calculates the approximate gradient of :math:`\sum_{i=1}^{n-1}f_i`.  For this approximation, every `update_frequency` number of iterations, a full gradient calculation is made at this "snapshot" point. Intermediate gradient calculations update this snapshot by taking a index :math:`i_k` and calculating the gradient of :math:`f_{i_k}`s at the current iterate and the snapshot, updating the approximate gradient to be:
+    
+        .. math ::
+            n*\nabla f_{i_k}(x_k) - n*\nabla f_{i_k}(\tilde{x}) + \nabla \sum_{i=0}^{n-1}f_i(\tilde{x}),
+            
+    where :math:`\tilde{x}` is the latest "snapshot" point and :math:`x_k` is the value at the current iteration. 
+    
+    Note
+    -----
+    Compared with the literature, we multiply by :math:`n`, the number of functions, so that we return an approximate gradient of the whole sum function and not an average gradient. 
+    
+    Reference
+    ---------
+    Johnson, R. and Zhang, T., 2013. Accelerating stochastic gradient descent using predictive variance reduction. Advances in neural information processing systems, 26.https://proceedings.neurips.cc/paper_files/paper/2013/file/ac1dd209cbcc5e5d1c6e28598e8cbbe8-Paper.pdf
+    
+
+    Parameters
+    ----------
+    functions : `list`  of functions
+        A list of functions: :code:`[f_{0}, f_{1}, ..., f_{n-1}]`. Each function is assumed to be smooth with an implemented :func:`~Function.gradient` method. All functions must have the same domain. The number of functions must be strictly greater than 1. 
+    sampler: An instance of a CIL Sampler class ( :meth:`~optimisation.utilities.sampler`) or of another class which has a `next` function implemented to output integers in {0, 1, ..., n-1}.
+        This sampler is called each time gradient is called and  sets the internal `function_num` passed to the `approximate_gradient` function.  Default is `Sampler.random_with_replacement(len(functions))`. 
+    update_frequency : int or None, optional
+        The frequency of updating the full gradient (taking a snapshot). The default is 2*len(functions) so a "snapshot" is taken every 2*len(functions) iterations. 
```

Comment:
Probably it is better to call the `update_frequency` to `snapshot_update_interval` and accept positive integers. 

I would suggest to use 0 in the case you want to force it never to update, although strictly speaking it would be infinite, rather than 0.